api/desecapi/serializers.py

```diff
@@ -56,23 +67,95 @@ def __call__(self, value):
         if isinstance(self.instance, Model) and value != getattr(self.instance, self.field_name):
             raise ValidationError(self.message, code='read-only-on-update')
 
-    def __repr__(self):
-        return '<%s>' % self.__class__.__name__
 
+class NonEmptyListOnCreate(Validator):
 
-class RRField(serializers.RelatedField):
-    def __init__(self, **kwargs):
-        kwargs['queryset'] = RR.objects.all()
-        super().__init__(**kwargs)
+    message = 'On creation, must be a non-empty list.'
+
+    def set_context(self, serializer_field):
+        """
+        This hook is called by the serializer instance,
+        prior to the validation call being made.
+        """
+        self.field_name = serializer_field.source_attrs[-1]
+        self.instance = getattr(serializer_field.parent, 'instance', None)
+
+    def __call__(self, value):
+        if self.instance is None:
+            if not isinstance(value, list) or len(value) == 0:
+                raise ValidationError(self.message, code='non-empty-list-on-creation')
+
+
+class StringField(CharField):
 
     def to_internal_value(self, data):
-        return dict(content=str(data))
+        return data
+
+    def run_validation(self, data=empty):
+        data = super().run_validation(data)
+        if not isinstance(data, str):
+            raise ValidationError('Must be a string.', code='must-be-a-string')
+        return data
+
+
+class RRsField(ListField):
+
+    def __init__(self, **kwargs):
+        super().__init__(child=StringField(), **kwargs)
+
+    def to_representation(self, data):
+        return [rr.content for rr in data.all()]
+
+
+class ConditionalExistenceModelSerializer(serializers.ModelSerializer):
+    """
+    Only considers data with certain condition as existing data.
+    If the existence condition does not hold, given instances are deleted, and no new instances are created,
+    respectively. Also, to_representation and data will return None.
+    Contrary, if the existence condition holds, the behavior is the same as DRF's ModelSerializer.
+    """
+
+    def existence_condition(self, arg):
+        """
+        Determine if arg is to be considered existing.
+        :param arg: Either a model instance or (possibly invalid!) data object.
+        :return: Whether we treat this as non-existing instance.
+        """
+        raise NotImplementedError
+
+    def to_representation(self, instance):
+        return None if not self.existence_condition(instance) else super().to_representation(instance)
+
+    @property
+    def data(self):
+        ret = super(Serializer, self).data  # note that we skip a level in the inheritance hierarchy
+        return None if ret is None else ReturnDict(ret, serializer=self)
+
+    def save(self, **kwargs):
+        validated_data = dict(
+            list(self.validated_data.items()) +
+            list(kwargs.items())
+        )
+
+        known_instance = bool(self.instance)
```

Comment:
Is this known to be reliable? The upstream `ModelSerializer` code uses `self.instance is not None`.

Reply:
I believe `self.instance` will never be a falsy value that is not `None`. (Empty lists, empty objects, `False` all do not seem to end up here.) However, no reason to change the default behavior; I changed it to model the upstream behavior.